### ISSUE TITLE
Mapping nullable Enum to custom object

### ISSUE THREAD

**Issue by WahidBitar** [MapsterMapper/Mapster#640]
for now I can have something like the following to register mapping for all enums in the system
```
config.NewConfig<Enum, KeyValueData>()
            .MapWith(s => new KeyValueData(s, Enums.ResourceManager));
```

However, it's not possible to have the same cofig for a nullable Enum. 

The following config is not working:

```
config.NewConfig<Enum?, KeyValueData?>()
            .MapWith(s => s == null ? null : new KeyValueData(s.ToString(), Enums.ResourceManager));
```

I'll get the following Exception

Error while compiling
source=Tests.FirstClass
destination=Tests.SecondClass
type=Map
-> InnerException:
Error while compiling
source=System.Nullable`1[Model.IdentityType]
destination=Model.KeyValueData
type=Map
--> InnerException:
No default constructor for type 'KeyValueData', please use 'ConstructUsing' or 'MapWith'

**Comment by WahidBitar** [comment]
I've figured a way out but still think it might be a better if we've somthing more helpful in the config. Especially from primitive types to classes. Anyways, my solution was simply like the following:

```
config.NewConfig(typeof(Nullable<>), typeof(KeyValueData))
            .IgnoreNullValues(true)
            .Map("key", "Value")
            .MapToConstructor(true);
```

**Comment by DocSvartz** [comment]
Hello @WahidBitar , if your goal is to skipped mapping when the Source is null, maybe you help [.IgnorIf()] or (https://github.com/MapsterMapper/Mapster/wiki/Ignoring-members#ignore-conditionally)

**Comment by DocSvartz** [comment]
Hello @WahidBitar, I ran into this problem while preparing a fix for the problem record detection. 

I think the problem is as follows: Nullable<T> in Mapster terminology should adapt as Record - forced creation via constructor. But now it happens as a Class where without special directives, there is only creation through a constructor without parameters.

**Comment by WahidBitar** [comment]
Thanks @DocSvartz , My main goal is not to skip the null values. My goal is to map from a nullable enum to my custom class `KeyValueData` in a general way.
I'll put here the entire aspects and my work around so you may take benifit of it.

Note: you may found something weard in my `KeyValueData` class because I have a general mapper from Enum to string because I put translation of the enums in the Resources files, so If you don't have this case your config should be simpler. Anyway here is my classes and config:

Mapping config:
```
public class GeneralMapper : IRegister
{
    public void Register(TypeAdapterConfig config)
    {
        // by default the following config will cause the Key property in the KeyValueData to get the value from the resources.
        // so I've updated the KeyValueData class to ignore the setter
        config.ForType<Enum, string?>()
            .MapWith(e => Enums.ResourceManager.GetString(e.ToResourceKey()) ?? e.ToString());
        
        config.NewConfig<Enum, KeyValueData?>()
            .MapWith(s => new KeyValueData(s));
        
        config.NewConfig(typeof(Nullable<>), typeof(KeyValueData))
            .MapToConstructor(true)
            .Map("key", "Value");
    }
}
```

the KeyValueData class:
```
public record KeyValueData
{
    private readonly string? keyHolder; // I had to set this field to this name becuase key will get the value from Resources
    private string? description;

    public KeyValueData(Enum key)
    {
        this.keyHolder = key.ToString();
        Description = Enums.ResourceManager.GetString(key.ToResourceKey()) ?? key.ToResourceKey();
    }

    public string Key
    {
        get => keyHolder!;
        set { } // Don't know why mapster call the setter after calling the constructor. I guess there is a bug related to this.
    }

    public string? Description
    {
        get => description;
        set => description ??= value;
    }
}
```

**Comment by DocSvartz** [comment]
@WahidBitar, I wanted to see the KeyValueData in order to better help you. Now I see next:

1) In this moment you Type KeyValueData has been detect Mapster as Class not  as Record. 
2) Activate Setter from `Property Key `possible linked from this bug in Class adapter I described it from this issue #633.

**Comment by DocSvartz** [comment]
Hello @WahidBitar , In version Mapster 7.4.0 and Mapster Tool  8.4.0  (Net 7.0) it working

from your last example: not Error 

```
 public class GeneralMapper : IRegister
 {
     public void Register(TypeAdapterConfig config)
     {
         // by default the following config will cause the Key property in the KeyValueData to get the value from the resources.
         // so I've updated the KeyValueData class to ignore the setter
         config.NewConfig<Enum?, KeyValueData?>()
             .MapWith(s => s == null ? null : new KeyValueData(s))
             .IgnoreNullValues(true).Compile();

     }
}

```
Result of Generation

```
public partial class FooMapper : TEstConsole.Mapster.IFooMapper
{
    public TEstConsole.clases.KeyValueData MapToEntity(System.Enum s)
    {
        return s == null ? null : new TEstConsole.clases.KeyValueData(s);
    }
}


```

**Comment by WahidBitar** [comment]
thank you so much @DocSvartz 
till now I don't use the Mapster Tool, but I am using the Mapster 7.4.0 and don't know why this config didn't work for me
```
config.NewConfig<Enum?, KeyValueData?>()
            .MapWith(s => s == null ? null : new KeyValueData(s.ToString(), Enums.ResourceManager));
```
so I ended up with my workaround solution

**Comment by DocSvartz** [comment]
The problem in here:
_source=System.Nullable`1[**Model.IdentityType**]
destination=Model.KeyValueData_
_new KeyValueData(s.ToString(), **Enums.ResourceManager**)_
Enums is Static or what is it ?

**Comment by DocSvartz** [comment]
Linked this #476

**Comment by DocSvartz** [comment]
@WahidBitar  @andrerav

Looks like I made a  fix from this Problem  
But there was a regression with this [test](https://github.com/DocSvartz/Mapster/blob/15ce5310a713e11741007b9e71e3880a997d0a66/src/Mapster.Tests/WhenProjecting.cs#L60)
 With Fix created Additional check for TSource:  `CreatedUserName = IIF((Param_0.CreatedUser == null), null, Param_0.CreatedUser.Name)`
I can't figure out if this is better or worse for Projection way Adapt ?

@andrerav i can integrated this fix from my PR  #634  or is it better to open a new PR?

**Comment by andrerav** [comment]
@DocSvartz Yes I prefer a separate PR :) Thanks for all your recent contributions, amazing work!

**Comment by stagep** [comment]
In order to identify issues that are still active, we are closing issues that we believe are either resolved or are dormant. If your issue is still active then please reopen. Thanks.
